Enabling HTTPS with local WebApp development (#1311)
microsoft/semantic-kernel: samples/apps/copilot-chat-app/webapp/README.md

See [../README.md](../README.md) for complete instructions on setting up and running the application.

## How to use HTTPS for local development

> If you need to run Copilot Chat with HTTPS, you need to create a certificate and sign it with a Certificate Authority (CA) that is trusted locally by your device and browser. You can do that easily with [mkcert](https://github.com/FiloSottile/mkcert#installation).

Reviewer: Nit: I think the following replacement would make things sound better:

If you need -> If you want

1. From an elevated shell, run
    ```

>     choco install mkcert

Reviewer 2: Introducing `choco` adds another Windows dependency? Users should be able to use this sample on WSL2/Linux (e.g. devcontainer) or on macOS as well in my opinion.

    ```
1. Create a new local CA by running
    ```
    mkcert -install
    ```

> 1. Create a new certificate for all hostnames you wish to run the app on. 

Reviewer: certificate for all -> certificate with all

> 1. Create a new certificate for all hostnames you wish to run the app on. 

Reviewer: certificate for all -> certificate with all

    > It is recommend you do this at the /WebApp directory level. 

    Run
    ```
    mkcert localhost 127.0.0.1 example.test
    ```
    You should see an output that looks like this:
    ```
    Created a new certificate valid for the following names 📜
        - "localhost"
        - "127.0.0.1"
        - "example.test"

    The certificate is at "./localhost+1.pem" and the key at "./localhost+1-key.pem" ✅

    It will expire on 1 September 2025
    ```

> 1. In the [WebApp .env](./.env) file, uncomment lines 11-13 and populate lines 12 and 13 with the certificate and key files generated in the step above.

Reviewer: Feel free to ignore this comment, but putting specific line numbers "condemns" you to update them later if the file changes...

    ```
    ...
    # To enable HTTPS, uncomment the following lines
    HTTPS="true"
    SSL_CRT_FILE=localhost+1.pem # Replace with your locally-trusted cert file
    SSL_KEY_FILE=localhost+1-key.pem # Replace with your locally-trusted cert key
    ```
1. In the [webapi appsettings.json](../webapi/appsettings.json) file, find the `"AllowedOrigins"` section, and add the URLs (with ports!) that you'll be running the apps on already with `https` prefixed. i.e., from the example hostnames above, the `"AllowedOrigins"` section should look like:
    ```
    ...
    // CORS
    "AllowedOrigins": [
        "http://localhost:3000", 
        "https://localhost:3000",
        "https://127.0.0.1:3000",
        "https://example.test:3000"
    ],
    ...
    ```
1. Add the same URLs (with ports!) as SPA redirect URIs to your AAD app registration. This can be done in Azure Portal ([https://portal.azure.com](https://ms.portal.azure.com/#home)).

> 1. Restart the `webapi` and `WebApp`, and Copilot Chat should be running locally with HTTPS!

Reviewer: I just have nits in this PR: the closing exclamation point makes it feel like it's insane that you're running in HTTPS.  :D  I would just keep it to a simple period.

Author: This made me chuckle, I changed it!

See [../README.md](../README.md) for complete instructions on setting up and running the application.

## How to use HTTPS for local development

> If you want to run Copilot Chat with HTTPS, you need to create a certificate and sign it with a Certificate Authority (CA) that is trusted locally by your device and browser. You can do that easily with [mkcert](https://github.com/FiloSottile/mkcert#installation).

Reviewer 3: I like the mkcert path you took here. Could we also add a note in this first part that includes a link for the Azure Key Vault certificate creation process? https://learn.microsoft.com/en-us/azure/key-vault/certificates/certificate-scenarios. I don't think we need much more than just a link to their page for the cert creation. They also have a section on creating an official globally accepted SSL certificate.

1. From an elevated shell, run

    ### Windows
    On Windows, use [Chocolatey](https://chocolatey.org/)
    ```
    choco install mkcert
    ```

    ### MacOS
    On macOS, use [Homebrew](https://brew.sh/)
    ```
    brew install mkcert
    ```

>     For installation on other OS and advanced topics (i.e., supported root stores), see the [official mkcert installation guide](https://github.com/FiloSottile/mkcert#installation). 

Reviewer 3: [nit] Could we explicitly call out Linux in this comment?
```suggestion
    For installation on Linux, other operating systems, and advanced topics (e.g., supported root stores), see the [official mkcert installation guide](https://github.com/FiloSottile/mkcert#installation). 
```

> 1. Create a new local CA by running

Reviewer 3: [nit] certificates are usually new for most people. I would recommend using as few acronyms as possible.
```suggestion
1. Create a new local certificate authority by running
```

    ```
    mkcert -install
    ```
1. Create a new certificate with all hostnames you wish to run the app on. 
    > It is recommend you do this at the /WebApp directory level. 

    Run
    ```
    mkcert localhost 127.0.0.1 example.test
    ```
    You should see an output that looks like this:
    ```
    Created a new certificate valid for the following names 📜
        - "localhost"
        - "127.0.0.1"
        - "example.test"

    The certificate is at "./localhost+1.pem" and the key at "./localhost+1-key.pem" ✅

    It will expire on 1 September 2025
    ```

    > **Warning**: (from the developers of mkcert) the rootCA-key.pem file that mkcert automatically generates gives complete power to intercept secure requests from your machine. Do not share it.
1. In the WebApp `.env` file, uncomment the following lines and populate with your respective certificate and key files generated in the step above.
    ```
    ...
    # To enable HTTPS, uncomment the following lines
    HTTPS="true"
    SSL_CRT_FILE=localhost+1.pem # Replace with your locally-trusted cert file
    SSL_KEY_FILE=localhost+1-key.pem # Replace with your locally-trusted cert key
    ```

> 1. In the [webapi appsettings.json](../webapi/appsettings.json) file, find the `"AllowedOrigins"` section, and add the URLs (with ports!) that you'll be running the apps on already with `https` prefixed. i.e., from the example hostnames above, the `"AllowedOrigins"` section should look like:

Reviewer 3: ```suggestion
1. In the [webapi appsettings.json](../webapi/appsettings.json) file, find the `"AllowedOrigins"` section, and add the URLs (with ports!) that you'll be running the apps on already with `https` prefixed. For instance, the `"AllowedOrigins"` section should look something like:
```

    ```
    ...
    // CORS
    "AllowedOrigins": [
        "http://localhost:3000", 
        "https://localhost:3000",
        "https://127.0.0.1:3000",
        "https://example.test:3000"
    ],
    ...
    ```

> 1. Add the same URLs (with ports!) as SPA redirect URIs to your AAD app registration. This can be done in Azure Portal ([https://portal.azure.com](https://ms.portal.azure.com/#home)).

Reviewer 3: ```suggestion
1. Add the same URLs (with ports!) as single-page authority (SPA) redirect URIs to your Azure Active Directory (AAD) application registration. This can be done in the [Azure Portal](https://portal.azure.com).
```

Reviewer: SPA stands for Single Page Application, not authority!

> 1. Restart the `webapi` and `WebApp`, and Copilot Chat should be running locally with HTTPS.

Reviewer 3: ```suggestion
1. Restart the `webapi` and `webapp` - Copilot Chat should be now running locally with HTTPS.
```
